### PR TITLE
fix: align various data structures for runtime efficiency

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: align various data structures for runtime efficiency

## Motivation and Context
data structures need to be aligned for better
runtime efficiency and usage, this can also
reduce the overall memory consumption over
a period of time.

## How to test this PR?
The tool used to align fields is fieldalignment from Go analysis repo

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation updated
- [ ] Unit tests added/updated

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#de63639c-1723-411a-bbfa-fdeee645336f) |
<a name="de63639c-1723-411a-bbfa-fdeee645336f"></a>
### 12045-85ffde2/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:30530
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp aaa443872323:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... FAILED in 33 seconds
{
  "name": "awscli",
  "duration": 2751,
  "function": "aws --endpoint-url http://minio-dev6.minio.io:30530 s3api copy-object --bucket awscli-mint-test-bucket-30676 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-30676/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 45 seconds
(7/15) Running minio-dotnet tests ... done in 45 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 12 seconds
{
  "args": {},
  "duration": 596,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 1 minutes and 31 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 141,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-9m8fq3, objectName = minio-java-test-2jp3j9l-copy, resource = /minio-java-test-9m8fq3/minio-java-test-2jp3j9l-copy, requestId = 16754D9E41B69026, hostId = e1ce4818-92bb-4e3e-a192-d0a2c754b929)\nrequest={method=PUT, url=http://minio-dev6.minio.io:30530/minio-java-test-9m8fq3/minio-java-test-2jp3j9l-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-fqd0t0/minio-java-test-2jp3j9l\nHost: minio-dev6.minio.io:30530\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210413T033623Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210413/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 416\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Tue, 13 Apr 2021 03:36:23 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 16754D9E41B69026\nX-Xss-Protection: 1; mode=block\nDate: Tue, 13 Apr 2021 03:36:23 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 48 seconds
(9/15) Running minio-py tests ... done in 3 minutes and 11 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "3376",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-10435/s3cmd-test-object-22583 193e220a-219d-465f-8a43-a0784bd32bc1\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 193e220a-219d-465f-8a43-a0784bd32bc1 put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 7 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#d5c9a8fa-9c95-4d4f-991e-625d07df126f) |
<a name="d5c9a8fa-9c95-4d4f-991e-625d07df126f"></a>
### 12045-912c922/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev6.minio.io:32281
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 3da3f7224175:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 42 seconds
(4/15) Running aws-sdk-ruby tests ... done in 4 seconds
(5/15) Running awscli tests ... FAILED in 32 seconds
{
  "name": "awscli",
  "duration": 2754,
  "function": "aws --endpoint-url http://minio-dev6.minio.io:32281 s3api copy-object --bucket awscli-mint-test-bucket-1046 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-1046/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 0 seconds
(6/15) Running mc tests ... done in 49 seconds
(7/15) Running minio-dotnet tests ... done in 41 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 29 seconds
{
  "args": {},
  "duration": 1032,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 1 minutes and 34 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 171,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-2qctlm, objectName = minio-java-test-2u2ko6t-copy, resource = /minio-java-test-2qctlm/minio-java-test-2u2ko6t-copy, requestId = 16754EAB82616F7A, hostId = 199a7d47-4413-4ba5-9756-609e7bcf567b)\nrequest={method=PUT, url=http://minio-dev6.minio.io:32281/minio-java-test-2qctlm/minio-java-test-2u2ko6t-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-haq8gj/minio-java-test-2u2ko6t\nHost: minio-dev6.minio.io:32281\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210413T035539Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210413/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 416\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Tue, 13 Apr 2021 03:55:39 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 16754EAB82616F7A\nX-Xss-Protection: 1; mode=block\nDate: Tue, 13 Apr 2021 03:55:39 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 48 seconds
(9/15) Running minio-py tests ... done in 3 minutes and 7 seconds
(10/15) Running s3cmd tests ... FAILED in 6 seconds
{
  "name": "s3cmd",
  "duration": "2991",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-12185/s3cmd-test-object-11565 90a7b5d5-f5b3-4253-a514-dc6e8417f83f\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 90a7b5d5-f5b3-4253-a514-dc6e8417f83f put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 6 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :x: [more...](#d9c2444a-e2af-41e9-b1c8-14ab2a14cdbb) |
<a name="d9c2444a-e2af-41e9-b1c8-14ab2a14cdbb"></a>
### 12045-21cd3bb/mint-compress-encrypt-dist-erasure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-c3.minio.io:30085
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 8546c80e161a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 1 seconds
(2/15) Running aws-sdk-java tests ... done in 1 seconds
(3/15) Running aws-sdk-php tests ... done in 44 seconds
(4/15) Running aws-sdk-ruby tests ... done in 3 seconds
(5/15) Running awscli tests ... FAILED in 31 seconds
{
  "name": "awscli",
  "duration": 2519,
  "function": "aws --endpoint-url http://minio-c3.minio.io:30085 s3api copy-object --bucket awscli-mint-test-bucket-31236 --key datafile-1-kB-copy --copy-source awscli-mint-test-bucket-31236/datafile-1-kB\n",
  "status": "FAIL",
  "error": "Hash mismatch expected 084e1383b70fb0c51acc680fef370023, got ac57de7156d7fc25ac1a65f81fa3989b"
}
(5/15) Running healthcheck tests ... done in 1 seconds
(6/15) Running mc tests ... done in 46 seconds
(7/15) Running minio-dotnet tests ... done in 34 seconds
(8/15) Running minio-go tests ... FAILED in 2 minutes and 9 seconds
{
  "args": {},
  "duration": 311,
  "error": "At least one of the pre-conditions you specified did not hold",
  "function": "CopyObjectPart(destination, source)",
  "message": "CopyObjectPart call failed",
  "name": "minio-go: testUnencryptedToSSES3CopyObjectPart",
  "status": "FAIL"
}
(8/15) Running minio-java tests ... FAILED in 1 minutes and 29 seconds
{
  "name": "minio-java",
  "function": "copyObject()",
  "args": "[match etag]",
  "duration": 234,
  "status": "FAIL",
  "error": "error occurred\nErrorResponse(code = PreconditionFailed, message = At least one of the pre-conditions you specified did not hold, bucketName = minio-java-test-17jebf3, objectName = minio-java-test-3cftk6f-copy, resource = /minio-java-test-17jebf3/minio-java-test-3cftk6f-copy, requestId = 16755068D416EBF2, hostId = daa00fe6-c1db-410a-8c47-7cbfc038ad7f)\nrequest={method=PUT, url=http://minio-c3.minio.io:30085/minio-java-test-17jebf3/minio-java-test-3cftk6f-copy, headers=x-amz-copy-source-if-match: 71cff0a060f852067e443ad1e24ae26c-1\nx-amz-copy-source: /minio-java-test-3v8ngua/minio-java-test-3cftk6f\nHost: minio-c3.minio.io:30085\nAccept-Encoding: identity\nUser-Agent: MinIO (Linux; amd64) minio-java/8.0.3\nContent-MD5: 1B2M2Y8AsgTpgAmY7PhCfg==\nx-amz-content-sha256: e3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855\nx-amz-date: 20210413T042732Z\nAuthorization: AWS4-HMAC-SHA256 Credential=*REDACTED*/20210413/us-east-1/s3/aws4_request, SignedHeaders=content-md5;host;x-amz-content-sha256;x-amz-copy-source;x-amz-copy-source-if-match;x-amz-date, Signature=*REDACTED*\n}\nresponse={code=412, headers=Accept-Ranges: bytes\nContent-Length: 418\nContent-Security-Policy: block-all-mixed-content\nContent-Type: application/xml\nETag: \"71cff0a060f852067e443ad1e24ae26c\"\nLast-Modified: Tue, 13 Apr 2021 04:27:32 GMT\nServer: MinIO\nVary: Origin\nX-Amz-Request-Id: 16755068D416EBF2\nX-Xss-Protection: 1; mode=block\nDate: Tue, 13 Apr 2021 04:27:32 GMT\n}\n >>> [io.minio.MinioClient.execute(MinioClient.java:775), io.minio.MinioClient.execute(MinioClient.java:563), io.minio.MinioClient.executePut(MinioClient.java:904), io.minio.MinioClient.copyObject(MinioClient.java:1232), FunctionalTest.testCopyObjectMatchETag(FunctionalTest.java:1850), FunctionalTest.copyObject(FunctionalTest.java:2016), FunctionalTest.runObjectTests(FunctionalTest.java:3757), FunctionalTest.runTests(FunctionalTest.java:3783), FunctionalTest.main(FunctionalTest.java:3927)]"
}
(8/15) Running minio-js tests ... done in 52 seconds
(9/15) Running minio-py tests ... done in 3 minutes and 0 seconds
(10/15) Running s3cmd tests ... FAILED in 5 seconds
{
  "name": "s3cmd",
  "duration": "2880",
  "function": "test_put_object_multipart",
  "status": "FAIL",
  "error": "WARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Retrying upload of /mint/data/datafile-65-MB\nWARNING: MD5 Sums don't match!\nWARNING: Too many failures. Giving up on '/mint/data/datafile-65-MB'\nERROR: \nUpload of '/mint/data/datafile-65-MB' part 1 failed. Use\n  /usr/local/bin/s3cmd abortmp s3://s3cmd-test-bucket-22935/s3cmd-test-object-13450 9350c3b9-5798-485d-9310-3b2a0852e4cd\nto abort the upload, or\n  /usr/local/bin/s3cmd --upload-id 9350c3b9-5798-485d-9310-3b2a0852e4cd put ...\nto continue the upload.\nERROR: Upload of '/mint/data/datafile-65-MB' failed too many times (Last reason: )"
}
(10/15) Running s3select tests ... done in 6 seconds
(11/15) Running security tests ... done in 0 seconds

Executed 11 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

File: cmd/bitrot.go [hunk 1]
```diff
@@ -78,13 +78,13 @@ func (a BitrotAlgorithm) String() string {
 
 // NewBitrotVerifier returns a new BitrotVerifier implementing the given algorithm.
 func NewBitrotVerifier(algorithm BitrotAlgorithm, checksum []byte) *BitrotVerifier {
-	return &BitrotVerifier{algorithm, checksum}
+	return &BitrotVerifier{checksum, algorithm}
```
Comment (klauspost) [inline cmd/bitrot.go:81]: Name values.

File: cmd/bitrot-whole.go [hunk 2]
```diff
@@ -90,7 +90,7 @@ func newWholeBitrotReader(disk StorageAPI, volume, filePath string, algo BitrotA
 		disk:       disk,
 		volume:     volume,
 		filePath:   filePath,
-		verifier:   &BitrotVerifier{algo, sum},
+		verifier:   &BitrotVerifier{sum, algo},
```
Comment (klauspost) [inline cmd/bitrot-whole.go:93]: Name values.

**Review by klauspost — APPROVED** [review]
I think most of these are only relevant on 32 bits, most of the changes are already aligned with 64 bit pointers, and some are just strange.

We can keep them, but please name the struct vars on init.

**Comment by harshavardhana** [comment]
Will close this and revisit this later..

**Pull request closed without merging**